### DISCUSSION TITLE
Does Strongswan Android 2.5.1 Support Post Quantum Configuration?

### DISCUSSION THREAD

**Discussion by oDinZu** [strongswan/strongswan#2159] [General]
I am setting up a Roadwarrior configuration with post quantum configuration and was wondering if Strongswan Android 2.5.1 supports post quantum configuration? 

I am new to the VPN stuff, but on launch of the app I see VPN types that use IKEv2; when I expand *advance settings*, I don't see any other options for using OQS algorithms..scrolling to the bottom of the advance settings, I do see an option that is for inputting custom IKEv2 algo's; I suppose I would manually add: `aes256-sha256-x25519-ke1_kyber3` there as an example.

I was reading through an article online and seen that their *local* Strongswan via **Linux** configuration also added **proposals** config with the OQS algorithm.

Edit: 

navigating through the https://docs.strongswan.org/docs/6.0/config/IKEv2CipherSuites.html has many options and configurations, however there isn't a list of options for post quantum algo's; would there be a page that lists these algo's for the configuration or how would I discover the *formatting* of these config options?

These are the algo's I compiled with 0.9.2, however their nomenclature is different from the nomenclature shared earlier.

`./swanctl --list-algs`

![Screenshot from 2024-03-07 19-38-21](https://github.com/strongswan/strongswan/assets/18726788/022849ea-fbb5-4a89-9800-14a3e150692a)

**Comment by tobiasbrunner** [comment]
> was wondering if Strongswan Android 2.5.1 supports post quantum configuration?

No.



> however there isn't a list of options for post quantum algo's;

There are no standardized algorithms for IKEv2 yet.

**Reply by oDinZu** [reply]
Alrighty, thank you for your reply @tobiasbrunner. 

Would it be possible to compile my own Android strongSwan with the OQS algorithms or is there a Android client that I am able to test?

The article I read detailed a site to site configuration that made use of a Linux client machine on Ubuntu. I assume that strongSwan Android application isn't much different from strongSwan Linux.

Edit: 

Upon further investigation, I should be able to do exactly the same thing on Arch Linux compile with OQS, but also for Android.

https://docs.strongswan.org/docs/6.0/os/androidVpnClientBuild.html

I would need to include: https://docs.strongswan.org/docs/5.9/plugins/kernel-libipsec.html plugin and add the OQS libraries, plus, maybe add another GUI option to the Android app for OQS or maybe just keep the GUI as it is and add it into the IKEv2 algo section as stated in previous conversation.

Edit: 

Also, they do have a script for building liboqs on Android, but I haven't tested it. 
https://github.com/open-quantum-safe/liboqs/blob/main/scripts/build-android.sh

I don't know if this would matter, but if we needed the cpp and java wrappers, they also have java-liboqs and cpp-liboqs.
https://github.com/orgs/open-quantum-safe/repositories?type=all

**Reply by oDinZu** [reply]
I integrated OQS 0.10.0 into Android 2.5.1, but I haven't tested anything yet..
![image](https://github.com/strongswan/strongswan/assets/18726788/52ffccb2-9d78-45fd-a230-2aa2185dea92)

**Reply by geedo0** [reply]
I don't have experience with the Android client, but you can find the names to use for your proposal [here](https://github.com/strongswan/strongswan/blob/six-beta/src/libstrongswan/crypto/key_exchange.c#L104). For example, I used `aes256gcm128-prfsha256-kyber3` for my own testing. If you can get liboqs to build for your target Android platform and load the oqs plugin into your strongSwan client it should work (assuming you have a similar responder on the other end that will accept your proposal).

**Comment by AlothNaveen** [comment]
@oDinZu @geedo0   Can you explain the steps, How you build Android client which support  Post Quantum Configuration....
// correct my steps and process
git clone --branch 0.8.0 https://github.com/open-quantum-safe/liboqs.git

cd liboqs
mkdir build
cd build
cmake -GNinja -DBUILD_SHARED_LIBS=ON -DCMAKE_INSTALL_PREFIX=<installation_path> ..
ninja
ninja install

cp -r <installation_path>/include/oqs /src/libstrongswan/plugins/oqs/
cp <installation_path>/lib/liboqs.so* /src/libstrongswan/plugins/oqs/

Edit src/libstrongswan/plugins/oqs/Makefile.am to link against the LibOQS shared library:
libstrongswan_oqs_la_LIBADD = -L$(top_srcdir)/src/libstrongswan/plugins/oqs -loqs

Enable the OQS plugin in the configuration step:
./configure --enable-oqs

Edit Android.mk to include the LibOQS shared library:
LOCAL_C_INCLUDES += $(LOCAL_PATH)/../../libstrongswan/plugins/oqs
LOCAL_SHARED_LIBRARIES += liboqs

building APK

These steps will enable Liboqs in APK? If not let me know ? it has to work for all architectures?
